src/algorithms/semihdp_sampler.cpp

```diff
@@ -0,0 +1,707 @@
+#include "semihdp_sampler.hpp"
+
+#include <proto/cpp/marginal_state.pb.h>
+
+#include <algorithm>
+#include <src/utils/eigen_utils.hpp>
+
+#include "../utils/distributions.hpp"
+
+SemiHdpSampler::SemiHdpSampler(const std::vector<Eigen::MatrixXd>& data,
+                               std::shared_ptr<BaseHierarchy> hier,
+                               bayesmix::SemiHdpParams params)
+    : data(data), params(params) {
+  ngroups = data.size();
+  n_by_group.resize(ngroups);
+  for (int i = 0; i < ngroups; i++) n_by_group[i] = data[i].size();
+  G0_master_hierarchy = hier->clone();
+  G00_master_hierarchy = hier->clone();
+  totalmass_rest = params.totalmass_rest();
+  totalmass_hdp = params.totalmass_hdp();
+}
+
+void SemiHdpSampler::initialize() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  omega = Eigen::VectorXd::Ones(ngroups).array();
+
+  int INIT_N_CLUS = 5;
+  Eigen::VectorXd probas = Eigen::VectorXd::Ones(INIT_N_CLUS);
+  probas /= (1.0 * INIT_N_CLUS);
+  rest_allocs.resize(ngroups);
+  is_used_rest.resize(ngroups);
+  table_allocs.resize(ngroups);
+  table_to_private.resize(ngroups);
+  table_to_shared.resize(ngroups);
+  rest_tables.resize(ngroups);
+  private_tables.resize(ngroups);
+  n_by_table.resize(ngroups);
+  rest_tables_pseudo.resize(ngroups);
+  n_by_table_pseudo.resize(ngroups);
+
+  for (int l = 0; l < INIT_N_CLUS; l++) {
+    std::shared_ptr<BaseHierarchy> hierarchy = G00_master_hierarchy->clone();
```

Comment:
This line should be outside the for loop (same with G0_master_hierarchy a few lines below)

Reply:
why? I need to clone new hierarchies every time

Reply:
Right, I forgot how cloning works, never mind